flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModel.java

```diff
@@ -0,0 +1,378 @@
+        }
+
+        /**
+         * @param input Input vector.
+         * @param topN Top N.
+         * @param priorityQueue Priority queue.
+         */
+        private void search(
+                DenseVector input,
+                Integer topN,
+                PriorityQueue<Tuple2<Double, Integer>> priorityQueue,
+                KnnModelData modelData) {
+            Tuple2<DenseVector, Double> sample = computeNorm(input);
+            Tuple2<Double, Integer> head = null;
+            for (int i = 0; i < modelData.getLength(); i++) {
+                List<Tuple2<Double, Integer>> values = computeDistance(sample, i);
+                for (Tuple2<Double, Integer> currentValue : values) {
+                    head = updateQueue(priorityQueue, topN, currentValue, head);
+                }
+            }
+        }
+
+        /**
+         * Updates queue.
+         *
+         * @param pq Queue.
+         * @param topN Top N.
+         * @param newValue New value.
+         * @param head Head value.
+         * @return Head value.
+         */
+        private <T> Tuple2<Double, T> updateQueue(
+                PriorityQueue<Tuple2<Double, T>> pq,
+                int topN,
+                Tuple2<Double, T> newValue,
+                Tuple2<Double, T> head) {
+            if (pq.size() < topN) {
+                pq.add(Tuple2.of(newValue.f0, newValue.f1));
+                head = pq.peek();
+            } else {
+                if (pq.comparator().compare(head, newValue) < 0) {
+                    Tuple2<Double, T> peek = pq.poll();
+                    assert peek != null;
+                    peek.f0 = newValue.f0;
+                    peek.f1 = newValue.f1;
+                    pq.add(peek);
+                    head = pq.peek();
+                }
+            }
+            return head;
+        }
+
+        /**
+         * Computes distance between sample and dictionary vectors.
+         *
+         * @param input Samples with l2 norm.
+         * @param index Dictionary vectors index.
+         * @return Distances.
+         */
+        private List<Tuple2<Double, Integer>> computeDistance(
+                Tuple2<DenseVector, Double> input, Integer index) {
+            Tuple3<DenseMatrix, DenseVector, int[]> data = modelData.getDictData().get(index);
+
+            DenseMatrix vectors = data.f0;
+            DenseMatrix distanceMatrix = new DenseMatrix(Objects.requireNonNull(vectors).numCols, 1);
+
+            DenseVector norm = data.f1;
+            double[] normL2Square = Objects.requireNonNull(norm).values;
+            BLAS.gemv(-2.0, vectors, true, input.f0, 0.0, new DenseVector(distanceMatrix.values));
+            for (int i = 0; i < distanceMatrix.values.length; i++) {
+                distanceMatrix.values[i] = Math.sqrt(Math.abs(distanceMatrix.values[i] + input.f1 + normL2Square[i]));
+            }
+
+            List<Tuple2<Double, Integer>> list = new ArrayList<>(0);
```

Comment:
nits: the initial capacity could be set to `curLabels.length` for better performance.

Reply:
OK